pkg/workspace/resource/node_claim.go

```diff
@@ -48,179 +49,203 @@ func NewNodeClaimManager(c client.Client, recorder record.EventRecorder, expecta
-		// Need to delete excess NodeClaims
-		nodesToDelete := currentNodeClaimCount - requiredNodeClaimsCount
-		klog.InfoS("Deleting excess NodeClaims",
-			"workspace", klog.KObj(wObj),
-			"current", currentNodeClaimCount,
-			"required", requiredNodeClaimsCount,
-			"toDelete", nodesToDelete)
+	}
 
+	if readyCount >= targetNodeCount {
+		// Enough NodeClaims are ready - update status condition to indicate success
+		if updateErr := workspace.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeNodeClaimStatus, metav1.ConditionTrue,
+			"NodeClaimsReady", fmt.Sprintf("Enough NodeClaims are ready (TargetNodeClaims: %d, CurrentReadyNodeClaims: %d)", targetNodeCount, readyCount)); updateErr != nil {
+			klog.ErrorS(updateErr, "failed to update NodeClaim status condition", "workspace", klog.KObj(wObj))
+			return false, fmt.Errorf("failed to update NodeClaim status condition(NodeClaimsReady): %w", updateErr)
+		}
+		return true, nil
+	} else {
 		if updateErr := workspace.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeNodeClaimStatus, metav1.ConditionFalse,
-			"DeletingNodeClaims", fmt.Sprintf("Deleting %d excess NodeClaims (current: %d, required: %d)", nodesToDelete, currentNodeClaimCount, requiredNodeClaimsCount)); updateErr != nil {
+			"NodeClaimNotReady", fmt.Sprintf("Ready NodeClaims are not enough (TargetNodeClaims: %d, CurrentReadyNodeClaims: %d)", targetNodeCount, readyCount)); updateErr != nil {
 			klog.ErrorS(updateErr, "failed to update NodeClaim status condition", "workspace", klog.KObj(wObj))
+			return false, fmt.Errorf("failed to update NodeClaim status condition(NodeClaimNotReady): %w", updateErr)
 		}
 
-		c.expectations.ExpectDeletions(c.logger, workspaceKey, nodesToDelete)
+		return false, nil
+	}
+}
 
-		// Sort NodeClaims for deletion: deletion timestamp set first, then not ready ones,
-		// then by creation timestamp (newest first)
-		sort.Slice(existingNodeClaims, func(i, j int) bool {
-			nodeClaimI := existingNodeClaims[i]
-			nodeClaimJ := existingNodeClaims[j]
+// ScaleDownNodeClaims scales down the NodeClaims for a workspace.
+// This function will be invoked after scaling down workloads of workspace in order to ensure nodes without pods can be deleted.
+func (c *NodeClaimManager) ScaleDownNodeClaims(ctx context.Context, wObj *kaitov1beta1.Workspace, existingNodeClaims []*karpenterv1.NodeClaim, nodesToDelete int) (bool, error) {
+	workspaceKey := client.ObjectKeyFromObject(wObj).String()
+	klog.InfoS("Scaling down excess NodeClaims", "workspace", workspaceKey, "toDelete", nodesToDelete)
 
-			deletingI := nodeClaimI.DeletionTimestamp != nil
-			deletingJ := nodeClaimJ.DeletionTimestamp != nil
-			if deletingI != deletingJ {
-				return deletingI // being deleted comes first
+	if nodesToDelete == 0 {
+		if curCondition := meta.FindStatusCondition(wObj.Status.Conditions, string(kaitov1beta1.ConditionTypeScalingDownStatus)); curCondition != nil {
```

Comment:
you don't need to check if the condition exists, `UpdateStatusConditionIfNotMatch` will assign the condition.